src/client/connect/http.rs

```diff
@@ -569,29 +569,30 @@ fn connect(
     connect_timeout: Option<Duration>,
 ) -> Result<impl Future<Output = Result<TcpStream, ConnectError>>, ConnectError> {
     // TODO(eliza): if Tokio's `TcpSocket` gains support for setting the
-    // keepalive timeout and send/recv buffer size, it would be nice to use that
-    // instead of socket2, and avoid the unsafe `into_raw_fd`/`from_raw_fd`
-    // dance...
+    // keepalive timeout, it would be nice to use that instead of socket2,
+    // and avoid the unsafe `into_raw_fd`/`from_raw_fd` dance...
     use socket2::{Domain, Protocol, Socket, Type};
+    use std::convert::TryInto;
+
     let domain = match *addr {
         SocketAddr::V4(_) => Domain::ipv4(),
         SocketAddr::V6(_) => Domain::ipv6(),
     };
     let socket = Socket::new(domain, Type::stream(), Some(Protocol::tcp()))
         .map_err(ConnectError::m("tcp open error"))?;
 
-    if config.reuse_address {
-        socket
-            .set_reuse_address(true)
-            .map_err(ConnectError::m("tcp set_reuse_address error"))?;
-    }
-
     // When constructing a Tokio `TcpSocket` from a raw fd/socket, the user is
     // responsible for ensuring O_NONBLOCK is set.
     socket
         .set_nonblocking(true)
         .map_err(ConnectError::m("tcp set_nonblocking error"))?;
 
+    if let Some(dur) = config.keep_alive_timeout {
+        socket
+            .set_keepalive(Some(dur))
+            .map_err(ConnectError::m("tcp set_keepalive error"))?;
+    }
+
```

Comment:
@seanmonstar should I just remove keepalive from the hyper API instead? That way, we could just drop the socket2 dep entirely.

Reply:
Humm.... *maybe*. It is a confusing property, but people do enable it to notice when a connection has died... :shrug:

Reply:
Yeah...it's looking like Tokio will probably not add support for more sophisticated configuration (e.g. keepalive retries/probe interval/etc). But it would be nice if it at least had an API for enabling/disabling keepalive and maybe setting the initial timeout before keepalive probes are sent, which should be more portable.